File: scripts/pdf-to-png.mjs
/**
 * Convert all PDF files in fish/binary_confidence_pdf/ to high-resolution PNGs
 * in fish/binary_confidence_png/
 *
 * Usage: node scripts/pdf-to-png.mjs
 */

import fs from 'fs';
import path from 'path';
import { createCanvas } from 'canvas';
import * as pdfjsLib from 'pdfjs-dist/legacy/build/pdf.mjs';

const BASE = path.resolve('fish');
const PDF_DIR = path.join(BASE, 'binary_confidence_pdf');
const PNG_DIR = path.join(BASE, 'binary_confidence_png');
const SCALE = 4; // 4x native resolution for crisp zoom

// Node canvas factory for pdfjs-dist
class NodeCanvasFactory {
  create(width, height) {
    const canvas = createCanvas(width, height);
    const context = canvas.getContext('2d');
    return { canvas, context };
  }
  reset(canvasAndContext, width, height) {
    canvasAndContext.canvas.width = width;
    canvasAndContext.canvas.height = height;
  }
  destroy(canvasAndContext) {
    canvasAndContext.canvas.width = 0;
    canvasAndContext.canvas.height = 0;
    canvasAndContext.canvas = null;
    canvasAndContext.context = null;
  }
}

const canvasFactory = new NodeCanvasFactory();

if (!fs.existsSync(PNG_DIR)) fs.mkdirSync(PNG_DIR, { recursive: true });

const files = fs.readdirSync(PDF_DIR).filter((f) => f.endsWith('.pdf'));
console.log(`Converting ${files.length} PDFs at ${SCALE}x scale...`);

let done = 0;
for (const file of files) {
  const pdfPath = path.join(PDF_DIR, file);
  const pngName = file.replace(/\.pdf$/i, '.png');
  const pngPath = path.join(PNG_DIR, pngName);

  try {
    const data = new Uint8Array(fs.readFileSync(pdfPath));
    const doc = await pdfjsLib.getDocument({ data, useSystemFonts: true, canvasFactory }).promise;
    const page = await doc.getPage(1);

    const viewport = page.getViewport({ scale: SCALE });
    const { canvas, context } = canvasFactory.create(
      Math.round(viewport.width),
      Math.round(viewport.height),
    );

    await page.render({ canvasContext: context, viewport }).promise;

    const pngBuffer = canvas.toBuffer('image/png');
    fs.writeFileSync(pngPath, pngBuffer);

    canvasFactory.destroy({ canvas, context });
    await doc.destroy();
    done++;
    if (done % 10 === 0 || done === files.length) {
      console.log(`  ${done}/${files.length}`);
    }
  } catch (err) {
    console.error(`  FAILED: ${file} — ${err.message}`);
  }
}

console.log(`Done. ${done} PNGs saved to ${PNG_DIR}`);

File: scripts/pdf-to-png.py
"""
Convert all PDF files in fish/binary_confidence_pdf/ to high-resolution PNGs.

Usage: python scripts/pdf-to-png.py
"""

import os
import fitz  # PyMuPDF

BASE = os.path.join(os.path.dirname(__file__), '..', 'fish')
PDF_DIR = os.path.join(BASE, 'binary_confidence_pdf')
PNG_DIR = os.path.join(BASE, 'binary_confidence_png')
SCALE = 4  # 4x native resolution

os.makedirs(PNG_DIR, exist_ok=True)

files = sorted(f for f in os.listdir(PDF_DIR) if f.lower().endswith('.pdf'))
print(f'Converting {len(files)} PDFs at {SCALE}x scale...')

done = 0
for f in files:
    pdf_path = os.path.join(PDF_DIR, f)
    png_path = os.path.join(PNG_DIR, f.rsplit('.', 1)[0] + '.png')

    try:
        doc = fitz.open(pdf_path)
        page = doc[0]
        mat = fitz.Matrix(SCALE, SCALE)
        pix = page.get_pixmap(matrix=mat, alpha=False)
        pix.save(png_path)
        doc.close()
        done += 1
        if done % 10 == 0 or done == len(files):
            print(f'  {done}/{len(files)}')
    except Exception as e:
        print(f'  FAILED: {f} — {e}')

print(f'Done. {done} PNGs saved to {PNG_DIR}')
